### PR TITLE
Alerting: Refactor ruleRoutine to take an entire ruleInfo instance

### PR DESCRIPTION
**Pull request by alexweav**
**What is this feature?**

Updates the rule routine to work in terms of the rule object, setting up for a future further simplification.
Updates tests to use the objects on the ruleInfo, making them more realistic.

**Why do we need this feature?**

n/a

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/ngalert/schedule/alert_rule.go [hunk 1]
```diff
@@ -58,3 +63,8 @@ func (a *alertRuleInfo) update(lastVersion ruleVersionAndPauseStatus) bool {
 		return false
 	}
 }
+
+// stop sends an instruction to the rule evaluation routine to shut down. an optional shutdown reason can be given.
+func (a *alertRuleInfo) stop(reason error) {
```
Comment (alexweav) [inline pkg/services/ngalert/schedule/alert_rule.go:68]: will eventually become an exported method on an exported interface

**Review by grobinson-grafana — COMMENTED** [review]
Looks fine, just one comment about something that surprised me!

File: pkg/services/ngalert/schedule/schedule.go [hunk 2]
```diff
@@ -345,8 +345,8 @@ func (sch *schedule) processTick(ctx context.Context, dispatcherGroup *errgroup.
 }
 
 //nolint:gocyclo
-func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertRuleKey, evalCh <-chan *evaluation, updateCh <-chan ruleVersionAndPauseStatus) error {
-	grafanaCtx = ngmodels.WithRuleKey(grafanaCtx, key)
+func (sch *schedule) ruleRoutine(key ngmodels.AlertRuleKey, ruleInfo *alertRuleInfo) error {
+	grafanaCtx := ngmodels.WithRuleKey(ruleInfo.ctx, key)
```
[the remark this answers is not in the document]
Comment (alexweav) [reply pkg/services/ngalert/schedule/schedule.go:349]: this is just a wrapper for a `context.WithValue` call. it just extends it.

[the remark this answers is not in the document]
Comment (grobinson-grafana) [reply pkg/services/ngalert/schedule/schedule.go:349]: Oh, duh! I misread it 😄 Thanks!

**Review by grobinson-grafana — APPROVED** [review]

**Review by JohnnyQQQQ — APPROVED** [review]

**Pull request merged**